packages/network-controller/tests/provider-api-tests/no-block-param.ts

```diff
@@ -0,0 +1,916 @@
+        );
+      });
+    });
+  }
+
+  it('throws a generic, undescriptive error if the request to the RPC endpoint returns a response that is not 405, 418, 429, 503, or 504', async () => {
+    await withMockedCommunications({ providerType }, async (comms) => {
+      const request = { method };
+
+      // The first time a block-cacheable request is made, the latest block
+      // number is retrieved through the block tracker first. It doesn't
+      // matter what this is — it's just used as a cache key.
+      comms.mockNextBlockTrackerRequest();
+      comms.mockRpcCall({
+        request,
+        response: {
+          id: 12345,
+          error: 'some error',
+          httpStatus: 420,
+        },
+      });
+      const promiseForResult = withNetworkClient(
+        { providerType },
+        async ({ makeRpcCall }) => makeRpcCall(request),
+      );
+
+      const errorMessage =
+        providerType === 'infura'
+          ? '{"id":12345,"jsonrpc":"2.0","error":"some error"}'
+          : "Non-200 status code: '420'";
+      await expect(promiseForResult).rejects.toThrow(errorMessage);
+    });
+  });
+
+  [503, 504].forEach((httpStatus) => {
+    it(`retries the request to the RPC endpoint up to 5 times if it returns a ${httpStatus} response, returning the successful result if there is one on the 5th try`, async () => {
+      await withMockedCommunications({ providerType }, async (comms) => {
+        const request = { method };
+
+        // The first time a block-cacheable request is made, the latest block
+        // number is retrieved through the block tracker first. It doesn't
+        // matter what this is — it's just used as a cache key.
+        comms.mockNextBlockTrackerRequest();
+        // Here we have the request fail for the first 4 tries, then succeed
+        // on the 5th try.
+        comms.mockRpcCall({
+          request,
+          response: {
+            error: 'Some error',
+            httpStatus,
+          },
+          times: 4,
+        });
+
+        comms.mockRpcCall({
+          request,
+          response: {
+            result: 'the result',
+            httpStatus: 200,
+          },
+        });
+        const result = await withNetworkClient(
+          { providerType },
+          async ({ makeRpcCall, clock }) => {
+            return await waitForPromiseToBeFulfilledAfterRunningAllTimers(
+              makeRpcCall(request),
+              clock,
+            );
+          },
+        );
+
+        expect(result).toStrictEqual('the result');
+      });
+    });
+
+    it(`causes a request to fail with a custom error if the request to the RPC endpoint returns a ${httpStatus} response 5 times in a row`, async () => {
+      await withMockedCommunications({ providerType }, (comms) => {
+        return returnsErrorAfterRetries(
+          providerType,
+          method,
+          'Some error',
+          buildFetchFailedErrorMessage(comms.rpcUrl, 'Some error'),
```

Comment:
It looks like this behavior differs from the extension. In the extension, this is split between the Infura client and generic JSON-RPC client; Infura responds with a "Gateway timeout" error and the other client returns an empty response.